fix(boot.go,src/healthsrv): add readiness and liveness probes

arschles opened the pull request:
Fixes #54

After this is merged and a new container is built, https://github.com/deis/charts/issues/86 needs to be done

Commit:
fix(boot.go,src/healthsrv): add health check server

and a single health check endpoint ‘/healthz’

Commit:
fix(boot.go): add log line to indicate starting the minio server

Commit:
fix(glide.lock,glide.yaml): add minio-go dependency

so that the health check server can use it to check the health of the
minio server

Commit:
fix(boot.go,src/healthsrv): use the minio client to test minio on health checks

Commit:
fix(src/healthsrv/healthz_handler.go): add logging for health check errors

Commit:
fix(Makefile): add ability to run go unit tests

Commit:
fix(boot.go,boot_test.go): move minio client creation logic out

and add a test for it

Commit:
fix(manifests/deis-minio-rc.yaml): add liveness and readiness probes

also specify the health check server port

bacongobbler commented on glide.lock:
```diff
@@ -229,7 +226,7 @@ imports:
   version: bb44bb2e4817fe71ba7082d351fd582e7d40e3ea
 - name: github.com/fsouza/go-dockerclient
-  version: 76fd6c68cf24c48ee6a2b25def997182a29f940e
+  version: ""
```
?

arschles replied:
I'm not sure why glide did this. Investigating...

arschles replied:
@bacongobbler after running `glide up` again (more precisely, `make glideup`), this version number didn't come back. In https://github.com/arschles/deis-minio/commit/08099c9503ea2908b00df85b422fd4eeee27dab8, however, another version number did come back for a launchpad.net dependency...

@technosophos suggested that I file an issue on glide for this, so I'm working on getting a solid set of repro steps down before I do.

bacongobbler commented on manifests/deis-minio-rc.yaml:
```diff
@@ -15,11 +15,27 @@ spec:
+          imagePullPolicy: Always
+          env:
+            - name: HEALTH_SERVER_PORT
+              value: "8095"
```
is there a specific reason this is 8095? Why not just use the default 8082 port or change the default to 8095 instead? :)

arschles replied:
Ah, thanks. I want to be explicit here, even if the specified port specified is the default. I'll fix the mismatch though!

Commit:
fix(manifests/deis-minio-rc.yaml): change the health check port to match the default

…while still keeping it explicitly defined

krancour commented:
I'm torn on whether it might be better to create the minio client used by the health check in / closer to the healthcheck code instead of doing so in `boot.go` and passing it in.  On the one hand, the complexity of creating that client seems like it should be handled a little closer to where it's used rather than the boot script which really handles overall orchestration.  On the other hand, secrets are used in constructing the minio client and you probably don't want to moves dependencies on kubernetes down into the healthcheck code.  Given that I can see both sides of this, I wouldn't let this stand in the way of my LGTM, but I just figured I'd mention it in case it inspires you in some way.

Commit:
fix(src/healthsrv/healtz_handler.go,src/storage): create a new interface for listing buckets

The Go minio client adheres to it, and it will be used for more easily
testing the healthz endpoint handler. The alternative was to use this
interface:
https://godoc.org/github.com/minio/minio-go#CloudStorageClient

…

Commit:
fix(src/healthsrv/healthz_handler_test.go): add test for the healthz handler

arschles commented:
@krancour good point, and thanks for that. I think the other option for minio client creation location is in the `Serve` function in the `./src/healthsrv` package. I don't think it would harm anything, but it seemed like it was more prudent to create it inside `boot.go`, since it may be used elsewhere in the future. Also, it may be beneficial for tests, in the case where I may want to test the `Serve` function in the future (I don't right now though since it's very basic)

Commit:
fix(glide.lock): re-run glide up

gets a version back for launchpad

mboersma commented:
This code looks good to me once the `glide` issue with go-dockerclient is sorted out.

Commit:
fix(glide.yaml): add version to minio-go dependency

Commit:
fix(glide.yaml,glide.lock): add utils and aboutme sub-packages

also re-run ‘glide up’, which adds versions back in for github
repositories.

finally, set the minio version to latest (as of this writing)

arschles commented:
@mboersma @bacongobbler after running a `glide up`, it looks like there are no more empty `version`s for github projects left

bacongobbler commented on glide.lock:
```diff
@@ -1,44 +1,37 @@
-  version: 68415e7123da32b07eab49c96d2c4d6158360e9b
-  repo: https://github.com/golang/protobuf
+  version: ""
+  repo: https://code.google.com/p/goprotobuf
```
code.google.com no longer exists. Any idea how this would occur?

arschles replied:
no idea. glide seems to toggle that just about every time I run `glide up`

bacongobbler commented on glide.lock:
```diff
@@ -1,44 +1,37 @@
-  version: ebbeafdc430e
 - name: code.google.com/p/go-uuid
-  version: 7dda39b2e7d5
+  version: ""
```
looks like github refs are fixed, but not for any references to code.google.com

arschles replied:
yup, but that's what I'd expect. did I misunderstand?

bacongobbler replied:
No I just realized this is the lockfile. Carry on :+1:

arschles replied:
k, thanks for looking again
